warehouse/macaroons/caveats/__init__.py

```diff
@@ -100,6 +101,38 @@ def verify(self, request: Request, context: Any, permission: str) -> Result:
         return Success()
 
 
+@as_caveat(tag=4)
+@dataclass(frozen=True)
+class OIDCProvider(Caveat):
+    oidc_provider_id: StrictStr
+
+    def verify(self, request: Request, context: Any, permission: str) -> Result:
+        # This caveat can only be used when the surrounding request is identified
+        # with OpenID, rather than a user or some other identity.
+        if not isinstance(request.identity, oidc_models.OIDCProvider):
+            return Failure(
+                "OIDC scoped token used outside of an OIDC identified request"
+            )
```

Comment:
Does this make sense? The identity comes from the token provided, not some characteristic of the surrounding request.

Reply:
I _think_ we need this because of Macaroons' ability to provide custom attenuation: a user could conceivably add a third-party `OIDCProvider` caveat to their otherwise valid user-identified macaroon, which would then fail because of the following ID comparison. But that comparison doesn't make sense to do in the first place, so having a check on the kind of requesting identity allows us to provide a slightly nicer error message.

(I believe that's the rationale for why we do a similar check in the `RequestUser` caveat.)

Reply:
Ah, I mostly meant the comment. The identity isn't coming from the 'surrounding request', it's coming from the identity we have correlated with this macaroon ID.

Reply:
Oh, I see what you mean! I can clarify the comment's intent.

Reply:
[b31c3b6](https://github.com/pypi/warehouse/pull/11272/commits/b31c3b625447dbf879d7b586de4c0ebdd89f989f)